### ISSUE TITLE
Polygons for iOS

### ISSUE THREAD

**Issue by d3mac123** [Yermo/nativescript-mapbox#204]
I would like to add airspace information to my current app (https://itunes.apple.com/us/app/ppg-finder/id1244768482?mt=8). Basically, draw circle areas showing the different airspaces - something like this https://dronelife.com/2017/03/14/where-can-i-fly-my-drone-these-5-apps-will-help/

or, even better, like this sample from Mapbox iOS documentation: https://www.mapbox.com/ios-sdk/examples/fill-pattern/

Any ETA for polygon implementation on iOS?

Thanks,
Alex

**Comment by andreMariano90** [comment]
Hi @d3mac123 . I don't own the project but you can always code it yourself in the right in the ios file.

I am developing an app and found myself in the same problem. Its actually pretty easy to code it following the Polyline sample. Here is where I got so far so it could meet my needs:
on the mapbox.ios.js
```javascript
    Mapbox.prototype.addPolygon = function (options, nativeMap) {
        return new Promise(function (resolve, reject) {
            var theMap = nativeMap || _mapbox.mapView;
            var points = options.points;
            if (points === undefined) {
                reject("Please set the 'points' parameter");
                return;
            }
            var coordinateArray = [];
            for (var p in points) {
                var point = points[p];
                coordinateArray.push([point.lng, point.lat]);
            }
            var polygonID = "polygon_" + options.id;
            if (theMap.style.sourceWithIdentifier(polygonID)) {
                reject("Remove the polyline with this id first with 'removePolylines': " + polygonID);
                return;
            }
            var geoJSON = "{\"type\": \"FeatureCollection\", \"features\": [{\"type\": \"Feature\",\"properties\": {},\"geometry\": {\"type\": \"Polygon\", \"coordinates\": [" + JSON.stringify(coordinateArray) + "]}}]}";
            var geoDataStr = NSString.stringWithString(geoJSON);
            var geoData = geoDataStr.dataUsingEncoding(NSUTF8StringEncoding);
            var geoDataBase64Enc = geoData.base64EncodedStringWithOptions(0);
            var geo = NSData.alloc().initWithBase64EncodedStringOptions(geoDataBase64Enc, null);
            var shape = MGLShape.shapeWithDataEncodingError(geo, NSUTF8StringEncoding);
            var source = MGLShapeSource.alloc().initWithIdentifierShapeOptions(polygonID, shape, null);
            theMap.style.addSource(source);

            var strokeLayer = MGLLineStyleLayer.alloc().initWithIdentifierSource(polygonID + "_stroke", source);
            strokeLayer.lineColor = MGLStyleValue.valueWithRawValue(!options.strokeColor ? UIColor.blackColor : (options.strokeColor instanceof color_1.Color ? options.strokeColor.ios : new color_1.Color(options.strokeColor).ios));
            strokeLayer.lineWidth = MGLStyleValue.valueWithRawValue(options.strokeWidth || 5);
            strokeLayer.lineOpacity = MGLStyleValue.valueWithRawValue(options.strokeOpacity === undefined ? 1 : options.strokeOpacity);
            theMap.style.addLayer(strokeLayer);

            var layer = MGLFillStyleLayer.alloc().initWithIdentifierSource(polygonID, source);
            layer.fillColor = MGLStyleValue.valueWithRawValue(!options.fillColor ? UIColor.blackColor : (options.fillColor instanceof color_1.Color ? options.fillColor.ios : new color_1.Color(options.fillColor).ios));
            layer.fillOpacity = MGLStyleValue.valueWithRawValue(options.fillOpacity === undefined ? 1 : options.fillOpacity);
            theMap.style.addLayer(layer);

            resolve();
        });
    };
```
As you can see... The MGLFillStyleLayer that is in the example by Mapbox is in this code. Just add anything else to meet your needs

**Comment by d3mac123** [comment]
Man, this is sweet! You should add it to Eddy's code!
Thank you very much!

**Comment by EddyVerbruggen** [comment]
This looks good, thanks! I'll integrate it myself.

**Comment by d3mac123** [comment]
@EddyVerbruggen Why did you remove this piece of the code provided by @andreMariano90? It is great to have the ability to set color and stroke to the polygons. Any chance you can add it back?

```
var strokeLayer = MGLLineStyleLayer.alloc().initWithIdentifierSource(polygonID + "_stroke", source);
            strokeLayer.lineColor = MGLStyleValue.valueWithRawValue(!options.strokeColor ? UIColor.blackColor : (options.strokeColor instanceof color_1.Color ? options.strokeColor.ios : new color_1.Color(options.strokeColor).ios));
            strokeLayer.lineWidth = MGLStyleValue.valueWithRawValue(options.strokeWidth || 5);
            strokeLayer.lineOpacity = MGLStyleValue.valueWithRawValue(options.strokeOpacity === undefined ? 1 : options.strokeOpacity);
            theMap.style.addLayer(strokeLayer);
```

**Comment by EddyVerbruggen** [comment]
I think because on Android it wasn't possible.

**Comment by andreMariano90** [comment]
Its the opposite actually. On Android the polygon accepts strokes and on the iOS it doesn't. This piece of code creates another line layer to simulate the stroke. But its actually another layer.

**Comment by EddyVerbruggen** [comment]
@andreMariano90 With the current Mapbox SDK you can no longer use `MGLStyleValue`, but I just found that you can do this:

```typescript
const strokeLayer = MGLLineStyleLayer.alloc().initWithIdentifierSource(polygonID + "_stroke", source);
strokeLayer.lineColor = NSExpression.expressionForConstantValue(!options.strokeColor ? UIColor.blackColor : (options.strokeColor instanceof Color ? options.strokeColor.ios : new Color(options.strokeColor).ios));
strokeLayer.lineWidth = NSExpression.expressionForConstantValue(options.strokeWidth || 5);
strokeLayer.lineOpacity = NSExpression.expressionForConstantValue(options.strokeOpacity === undefined ? 1 : options.strokeOpacity);
theMap.style.addLayer(strokeLayer);
```

And on Android this (width and opacity are not supported unfortunately):

```typescript
polygonOptions.strokeColor(Mapbox.getAndroidColor(options.strokeColor));
```

**Comment by d3mac123** [comment]
I've tried the above code and it did not render the stroke line...

**Comment by EddyVerbruggen** [comment]
@d3mac123 iOS or Android?

**Comment by d3mac123** [comment]
ios

**Comment by EddyVerbruggen** [comment]
Don't worry, I'm working on adding it to the plugin and iOS will be fine (it's working for me). On Android (I now recall) I run into this issue when setting the stroke color (it's barely visible): https://github.com/mapbox/mapbox-gl-native/issues/5676

**Comment by d3mac123** [comment]
If you can add this line (iOS plugin) in the setCenter function, it would be great (the map follow the user):
`theMap.userTrackingMode = MGLUserTrackingMode.MGLUserTrackingModeFollowWithCourse;`

As I have today:
```
    Mapbox.prototype.setCenter = function (options, nativeMap) {
        return new Promise(function (resolve, reject) {
            try {
                var theMap = nativeMap || _mapbox.mapView;
                var animated = options.animated === undefined || options.animated;
                var coordinate = CLLocationCoordinate2DMake(options.lat, options.lng);
                theMap.userTrackingMode = MGLUserTrackingMode.MGLUserTrackingModeFollowWithCourse;
                theMap.setCenterCoordinateAnimated(coordinate, animated);
                resolve();
            }
            catch (ex) {
                console.log("Error in mapbox.setCenter: " + ex);
                reject(ex);
            }
        });
    };
```

**Comment by EddyVerbruggen** [comment]
That seems a bit off topic..

**Comment by d3mac123** [comment]
Sorry about that. This is related to this case https://github.com/EddyVerbruggen/nativescript-mapbox/issues/144 (just mentioned because you will be working on the code and this is a simple line to add - in my case I have to add it manually every time you update the plugin).

**Comment by EddyVerbruggen** [comment]
Yeah but this is a closed issue and I did the work in #235. Let's continue in #144, thx!
